addons/hr_timesheet/project_timesheet.py

```diff
@@ -1,69 +1,35 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from openerp.osv import fields, osv
+from openerp import models, fields, api
 
 
-class task(osv.osv):
+class Task(models.Model):
     _inherit = "project.task"
 
-    # Compute: effective_hours, total_hours, progress
-    def _hours_get(self, cr, uid, ids, field_names, args, context=None):
-        res = {}
-        for task in self.browse(cr, uid, ids, context=context):
-            res[task.id] = {'effective_hours': 0.0, 'remaining_hours': task.planned_hours, 'progress': 0.0, 'total_hours': task.planned_hours, 'delay_hours': 0.0}
-        tasks_data = self.pool['account.analytic.line'].read_group(cr, uid, [('task_id', 'in', ids)], ['task_id','unit_amount'], ['task_id'], context=context)
-        for data in tasks_data:
-            task = self.browse(cr, uid, data['task_id'][0], context=context)
-            res[data['task_id'][0]] = {'effective_hours': data.get('unit_amount', 0.0), 'remaining_hours': task.planned_hours - data.get('unit_amount', 0.0)}
-            res[data['task_id'][0]]['total_hours'] = res[data['task_id'][0]]['remaining_hours'] + data.get('unit_amount', 0.0)
-            res[data['task_id'][0]]['delay_hours'] = res[data['task_id'][0]]['total_hours'] - task.planned_hours
-            res[data['task_id'][0]]['progress'] = 0.0
-            if (task.planned_hours > 0.0 and data.get('unit_amount', 0.0)):
-                res[data['task_id'][0]]['progress'] = round(min(100.0 * data.get('unit_amount', 0.0) / task.planned_hours, 99.99),2)
-            # TDE CHECK: if task.state in ('done','cancelled'):
-            if task.stage_id and task.stage_id.fold:
-                res[data['task_id'][0]]['progress'] = 100.0
-        return res
-
-    def _get_task(self, cr, uid, id, context=None):
-        res = []
-        for line in self.pool.get('account.analytic.line').search_read(cr,uid,[('task_id', '!=', False),('id','in',id)], context=context):
-            res.append(line['task_id'][0])
-        return res
-
     def _get_total_hours(self):
-        return super(task, self)._get_total_hours() + self.effective_hours
+        return super(Task, self)._get_total_hours() + self.effective_hours
+
+    @api.depends('timesheet_ids', 'timesheet_ids.unit_amount', 'planned_hours')
+    def _hours_get(self):
+        timesheet_lines = self.env['account.analytic.line'].read_group([('task_id', 'in', self.ids)], ['task_id', 'unit_amount'], ['task_id'])
+        time_per_task = dict((data['task_id'][0], data['unit_amount']) for data in timesheet_lines)
+        for task in self:
+            task.effective_hours = time_per_task.get(task.id, 0.0)
+            task.remaining_hours = task.planned_hours - time_per_task.get(task.id, 0.0)
+            task.total_hours = task.planned_hours
+            task.delay_hours = task.remaining_hours * -1
```

Comment:
I updated the computation of delay_hours. It's not super useful, but at least it's not always 0. 
That being said, it only appears in one report, so I'm not convinced that there is a need to keep this field.